### DISCUSSION TITLE
Helping R-Instat run more efficiently through better filling of the user controls

### DISCUSSION THREAD

**Discussion by rdstern** [IDEMSInternational/R-Instat#6329] [Ideas]
As we expect to distribute R-Instat more widely, we should make sure - as far as possible - that it works reasonably.

I have been considering the maximum number of variables in a data frame and also the maximum number of levels in a factor and they are a bit frightening for our current R-Instat.  

Variables first.  What are the limits.  For reference I give them in spreadsheet software as well as R:
a) Excel as xls file:  256  (2<sup>8</sup>) columns and 65,536 (2<sup>16</sup>) rows.
b) Excel now is 1,048,576 rows (2<sup>20</sup>) and 16,384 columns (2<sup>14</sup>) 
c) LibreOffice is also 1,048,576 rows (2<sup>20</sup>), 1024 columns (2<sup>10</sup>) and 1,073,741,824 (2<sup>30</sup>)  - that's one billion - cells
d) In SPSS the maximum number of rows is unlimited on the 64 bit versions, 2,147,483,647 (that's 2<sup>31</sup> - 1) on the 32 bit.  The limit on number of variables is 2,147,483,647 (that's 2<sup>31</sup> - 1)

Now to R:  
An R data frame can have 2,147,483,647 (that's 2<sup>31</sup> - 1) rows of data and/or also the same number of columns - variables!

Suddenly this has frightened me!  I give an example:

The diamonds data is used in Tutorial 1.  It has 53 thousand rows and 10 columns.  The have a transpose function in R-Instat.  This would make it into 10 rows and 53 thousand columns (variables).
R will be quite happy (Excel would not cope).  R-Instat should do it, but could it open a dialogue and fill the data selector with 53000 entries?

We therefore need to do something about the data selector.  I suggest it is urgent, simply because sceptics about click and point systems are likely to try things like that.  We should ensure that our dialogues don't break down.  

So, I suggest an urgent action is to do something about the data selector.  For now it may be that we simply impose a maximum.  You can still use R-Instat - scripts - or send to RStudio - but you can't use the dialogues.  

Even then, we should ensure it is reasonable to use up to that maximum.  The most I have seen in a data set for analysis is 4000 variables in one from SPSS in Kenya.

Maybe we can do better.  After all SPSS is menu driven and allows the same number of variables as R.

There are a few other dialogue that can have a problem, e.g. Reorder columns.  

Now the second problem.  Where R wins over SPSS (in my view anyway!) is its explicit use of factors.  However, there is also no particular limit to the number of levels of a factor.  And we have some dialogues that show the levels of a factor.  In particular there is the levels/labels dialogue, and also the filter, when we filter on the levels of a factor.

One of the responses in a discussion on this says:
![image](https://user-images.githubusercontent.com/11226469/112096049-7ab4f680-8b95-11eb-8739-5bf1586e2669.png)

I have put this as 0.7.x, but would like some discussion on this topic at least, within the sprint.  I suggest a deliverable could be a comment below on what we propose to do.  Further, I suggest doing something (even if simplified) is a job for April/May.

**Comment by dannyparsons** [comment]
Happy with a discussion on this before the end of the sprint. Since you said it's a discussion, you've given me the go-ahead to covert it to a "discussion"!

**Comment by rdstern** [comment]
I now have a possible plan, because it would be great to be able to still use R-Instat in an ordinary way, when we have a very odd data set.  And that could be something we discuss openly.

There are 2 problems with the dialogues - not with R itself!  The first is when a factor has lots of levels (how many is lots?).  That's because we often fill dialogues with each level of a factor.  The second is when there are many variables, because the data selector fills with all the variables.

My solution is that we have limits for each, just for dialogues.  I'm a bit clearer for factors where we have improved the commands recently (still to be merged as I add this).  So my plan uses the Other and/or Lump options for factors.  So we have a default lump  in our data options, which the user can change.  So, when you fill the dialogue - if there are more than our set number of levels, then it only shows the lumped data, i.e. it has an Other category in the dialogue, even if you have not done that (yet) in the data.  That's a whole group of levels.  Perhaps you can select them, etc.

Same for variables, now we have select.  We become able to display a whole group of  variables, with their select names.  If the dataset is very wide then we display (say) x00 and the rest as a group.  Much better is to define the groups yourself and use those instead when filling the selector.

I am really happy with the factor solution!  For the variable solution I would like also to fill the selector efficiently, so this is only needed in extreme circumstances.  I would also like obvious commands that produce very wide datasets to offer a select option.  An example is the Transpose dialogue.
